### ISSUE TITLE
Distributed hyper parameter optimization for dask.

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#6525]
Right now, integration between dask and various single node machine learning libraries are implemented as standalone dask extensions like dask-ml and dask-optuna.  These can be used with xgboost when xgboost is performing single node training.  That's using `XGBRegressor` and friends with them, instead of using `xgboost.dask.DaskXGBRegressor`.  If users want to train the entire dataset on 1 model, the dask interface is required.  The underlying issue is xgboost by itself is a distributed learning library employing a MPI like communication framework, but those extensions are designed to extend single node libraries.  To resolve it, we need to design python wrappers that can glue them together.

Optuna is an exception as it's using callback function in xgboost, so the xgboost.dask interface can be adopted to optuna.  I will submit some changes with demos later.  Others like grid searching are more difficult to implement.

Related: https://github.com/dmlc/xgboost/issues/5347

cc @pseudotensor @sandys

**Comment by sandys** [comment]
Thank you for taking this forward.

We are also seriously looking at Ray Distributed - because of its native
Integration with kubernetes (
https://docs.ray.io/en/master/cluster/kubernetes.html ) and its success at
Ant Financial in China (
https://docs.ray.io/en/master/cluster/kubernetes.html )


My thought is that you won't be able to make it generic enough to fit all
frameworks. And probably it's an overkill of engineering to do so.

The one thing I would strongly request you to align towards is kubernetes
compatibility. Because it's the management framework that has massive
amount of support.

It is for this reason that Torch chose to write a thin wrapper over
kubernetes for its distributed training library -
https://pytorch.org/elastic/0.2.1/index.html

https://aws.amazon.com/blogs/containers/fault-tolerant-distributed-machine-learning-training-with-the-torchelastic-controller-for-kubernetes/

So I would suggest when you look at solving this problem from a kubernetes
first standpoint. Compatibility with every framework is going to take up
more of your time than you might be willing to spend.

Sincerely
Sandeep

**Comment by trivialfis** [comment]
@sandys k8s support is in 1.3.

**Comment by trivialfis** [comment]
@sandys Use https://kubernetes.dask.org/en/latest/ to create the cluster, and train xgboost dask as usual.

**Comment by sandys** [comment]
Yes. Although dask has some independent issues with kubernetes - on the
scaling side.

Ray uses kubernetes scheduler. Dask brings a centralised scheduler of its
own.

In any case, I would strongly suggest Xgboost pick one framework and play
well with kubernetes. Rather than expose bindings for many many frameworks
to work with Xgboost.

Just my 0.02$

**Comment by trivialfis** [comment]
@sandys Thanks for your feedback.  k8s is definitely important to us.

**Comment by CarterFendley** [comment]
Hi @trivialfis thanks for your work on this. Just reading these threads now.

I see in the [other issue](https://github.com/dmlc/xgboost/issues/5347#issuecomment-600031975) you recommend using `sklearn` optimizers for out-of-core datasets and `dask_ml` optimizers for datasets which can fit in memory. I am wondering if you can expand / explain more on why `xgboost.dask` is incompatible with `dask_ml` at the moment.

What would it take to, in your words, "design python wrappers that can glue them together."?

**Comment by trivialfis** [comment]
Hi @CarterFendley , there is some on-going work on HOP, could you please take a look at https://github.com/coiled/dask-xgboost-nyctaxi and see if it helps?
